client/js/analysis/gdpr.js

```diff
@@ -11,9 +11,63 @@
  * track was listened.
  */
 
+/**
+ * @typedef {object} GDPRRecord
+ * @property {Date} endTime Date and time of when the stream ended in UTC
+ * format.
+ * @property {string} artistName Name of "creator" for each stream (e.g. the
+ * artist name if a music track).
+ * @property {string} trackName Name of items listened to or watched (e.g. title
+ * of music track or name of video).
+ * @property {number} msPlayed How many mili-seconds the
+ * track was listened.
+ */
+
 // global variables are not available in ES modules
 const {zip} = window;
 
+/**
+ * Gets the value associated with `key` from `map`. If `key` is not in `map`,
+ * then `fn` will be called and the value it returns will be placed in the map
+ * and returned from this function.
+ *
+ * @template K
+ * @template V
+ * @param {Map<K, V>} map The map to retrieve a value from.
+ * @param {K} key The key associated with the value to retrieve.
+ * @param {(key: K) => V} fn Generates a default value if the key was not in the
+ * map.
+ * @returns {V} The value associated with `key` in `map`, or the return value of
+ * `fn` if there was no such value.
+ */
+function getOrDefault(map, key, fn) {
+  let value = map.get(key);
+  if (value === undefined) {
+    value = fn(key);
+    map.set(key, value);
+  }
+  return value;
+}
+
+/**
+ * Updates the value associated with `key` in `map`. If `key` is not in `map`,
+ * then `fn` will be called with `undefined` as the value.
+ *
+ * @template K
+ * @template V
+ * @param {Map<K, V>} map The map to update.
+ * @param {K} key The key associated with the value to update.
+ * @param {(value: V | undefined, key: K) => V} fn Returns the new value to be
+ * associated with `key` in `map`.
+ * @returns {Map<K, V>} `map`
+ */
+function update(map, key, fn) {
+  const value = map.get(key);
+  map.set(key, fn(value, key));
```

Comment:
I'm assuming the `fn(value, key)` you put here handles the key in map case as well as key is not in map. If that's the case, then why do you need the if statement in the `getOrDefault()` function?

Reply:
makes sense; i wrote these functions and then deleted them and then put them back b/c i wasn't sure whether they were actually necessary or not. i'll just get rid of them